File: linear_search.cc
#include <iostream>
#include <vector>
#include <chrono>

void usage() {
   std::cerr << "usage: ./linear_search <set_length>" << std::endl;
   std::cerr << "reads from stdin <set_lenght> rows formated as: x n array[n]" << std::endl;
   std::cerr << "where \"x\" is a value in the array and \"array\" is an array with lenght n" << std::endl;
   exit(1);
}

void read_vector(std::vector<int> &v) {
   int n = v.size();
   for (int i = 0; i < n; ++i) std::cin >> v[i];
}

int linear_search(int x, const std::vector<int> &v) {
   int n = v.size();
   for (int i = 0; i < n; ++i) {
      if (v[i] == x) return i;
   }
   return -1; // never reached because precondition: x exists in v.
}

int main(int argc, char **argv) {
   if (argc != 2) usage();
   
   int set_length = atoi(argv[1]);
   // mark start time
   std::cout << "N TIME" << std::endl; // data set header
   for (int i = 0; i < set_length; ++i) {
      int x, n;
      std::cin >> x >> n; // catch number to find.
      std::vector<int> v(n);
      read_vector(v);
      
      auto start = std::chrono::high_resolution_clock::now();
      linear_search(x, v);
      // mark end time
      auto end = std::chrono::high_resolution_clock::now();
      // calc delta.
      auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
      // print delta on stdout.
      std::cout << n << ' ' << duration.count() << std::endl; // time in nanoseconds
   }
}

File: linear_search.py
import sys
import time

def usage():
   print("usage: python linear_search.py <set_length>", file=sys.stderr)
   print("reads from stdin <set_length> rows formatted as: x n array[n]", file=sys.stderr)
   print("where 'x' is a value in the array and 'array' is an array with length 'n'", file=sys.stderr)
   sys.exit(1)

def read_vector(line):
   parts = list(map(int, line.split()))
   x = parts[0]   # First value is the number to find
   n = parts[1]   # Second value is the length of the vector.
   if len(parts) != n + 1:
      print(f"Error: Expected {n + 1} values, got {len(parts)}", file=sys.stderr)
      sys.exit(1)
   v = parts[2:]  # Remaining values are the vector
   return x, n, v

def linear_search(x, v):
   for i, val in enumerate(v):
      if val == x:
         return i
   return -1  # never reached because precondition: x exists in v.

def main():
   if len(sys.argv) != 2:
      usage()

   try:
      set_length = int(sys.argv[1])
   except ValueError:
      usage()

   for _ in range(set_length):
      line = input().strip()  # Read the entire line
      x, n, v = read_vector(line)
      
      start = time.perf_counter_ns()  # mark start time
      linear_search(x, v)
      end = time.perf_counter_ns()  # mark end time

      # calc delta and print elapsed time
      duration = end - start
      print(f"{duration}") # time in nanoseconds

if __name__ == "__main__":
   main()
